File: src/main/java/dev/rafi/vitelist/database/DataSource.java
package dev.rafi.vitelist.database;

import com.j256.ormlite.dao.Dao;
import com.j256.ormlite.dao.DaoManager;
import com.j256.ormlite.jdbc.JdbcConnectionSource;
import com.j256.ormlite.support.ConnectionSource;
import com.j256.ormlite.table.TableUtils;
import dev.rafi.vitelist.ViteList;
import dev.rafi.vitelist.database.model.ServerModel;
import dev.rafi.vitelist.storage.Config;
import dev.rafi.vitelist.utils.QueryUtils;
import lombok.Getter;

import java.io.File;
import java.io.IOException;
import java.sql.SQLException;

public class DataSource {
    @Getter private static Dao<ServerModel, String> serverDao;
    @Getter private ConnectionSource connectionSource;

    public DataSource() throws IOException, SQLException {
        if (Config.TYPE.equalsIgnoreCase("MySQL")) {
            MySQL();
        } else {
            SQLite();
        }
    }


    private void SQLite() throws IOException, SQLException {
        File file = new File(ViteList.getDataDirectory().toUri().getPath().replace("vitelist", "ViteList") + "/" + "data.db");
        if (!(file.exists())) file.createNewFile();

        String databaseUrl = "jdbc:sqlite:" + ViteList.getDataDirectory().toUri().getPath() + "/data.db";

        connectionSource = new JdbcConnectionSource(databaseUrl);

        setupTables();
    }

    private void MySQL() throws IOException, SQLException {
        String databaseUrl = "jdbc:mysql://" + Config.HOST + ":" + Config.PORT + "/" + Config.NAME + "?useSSL=false&autoReconnect=true";

        connectionSource = new JdbcConnectionSource(databaseUrl, Config.USER, Config.PASSWORD);

        setupTables();
    }

    private void setupTables() throws SQLException {
        serverDao = DaoManager.createDao(connectionSource, ServerModel.class);
        TableUtils.createTableIfNotExists(connectionSource, ServerModel.class);
    }
}

File: src/main/kotlin/dev/rafi/vitelist/database/DataSource.kt
package dev.rafi.vitelist.database

import dev.rafi.vitelist.ViteList
import dev.rafi.vitelist.database.model.Server
import dev.rafi.vitelist.storage.Config
import org.jetbrains.exposed.sql.Database
import org.jetbrains.exposed.sql.SchemaUtils
import org.jetbrains.exposed.sql.transactions.TransactionManager
import org.jetbrains.exposed.sql.transactions.transaction
import java.io.File
import java.sql.Connection

class DataSource {

    init {
        initDataBase()

    }

    private fun initDataBase() {
        val dbType = Config.DB_TYPE
        val sqLite = "SQLite"
        val mySql = "MySQL"

        if (dbType.equals(sqLite, true)) {
            sqLite()
        } else if (dbType.equals(mySql, true)) {
            mySql()
        } else {
            sqLite()
        }

        transaction {
            SchemaUtils.create(Server)
        }
    }

    private fun sqLite() {
        val sqLiteDriver = "org.sqlite.JDBC"
        val dbFile = File(ViteList.getDataDir().toFile(), "data.db")

        if (!(dbFile.exists())) dbFile.createNewFile()

        Database.connect("jdbc:sqlite:${dbFile.toPath()}", sqLiteDriver)
        TransactionManager.manager.defaultIsolationLevel = Connection.TRANSACTION_SERIALIZABLE
    }

    private fun mySql() {
        val mySqlDriver = "com.mysql.cj.jdbc.Driver"

        Database.connect("jdbc:mysql://${Config.DB_HOST}:${Config.DB_PORT}/${Config.DB_NAME}", driver = mySqlDriver, user = Config.DB_USERNAME, password = Config.DB_PASSWORD)
    }
}